### PR TITLE
Test/implement more aspects of primary constructor parameter capturing.

### PR DESCRIPTION
**Pull request by AlekseyTs**
- Nullable analysis
- Region analysis
- Struct layout cycles
- Misc.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedPrimaryConstructorParameterBackingFieldSymbol.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        protected override IAttributeTargetSymbol AttributeOwner
+            => this;
+
+        internal override Location ErrorLocation
+            => ParameterSymbol.Locations.FirstOrDefault() ?? NoLocation.Singleton;
+
+        protected override SyntaxList<AttributeListSyntax> AttributeDeclarationSyntaxList
+            => default;
+
+        public override Symbol? AssociatedSymbol
```
Comment (jjonescz) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedPrimaryConstructorParameterBackingFieldSymbol.cs:38]: >AssociatedSymbol [](http://example.com/codeflow?start=32&length=16)

Should this return the `ParameterSymbol`? The documentation says "If this field serves as a backing variable for an automatically generated property or a field-like event, returns that property/event. Otherwise returns null. Note, the set of possible associated symbols might be expanded in the future to reflect changes in the languages."

**Comment by jjonescz** [comment]
>             return ((SourceMemberContainerTypeSymbol)ContainingType).IsNullableEnabledForConstructorsAndInitializers(IsStatic);

The cast is redundant now.

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordConstructor.cs:58 in df89318. [](commit_id = df89318d175c2c0f3768e7d5671469d7de975856, deletion_comment = False)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 2]
```diff
@@ -3222,6 +3199,128 @@ ImmutableArray<SynthesizedSimpleProgramEntryPointSymbol> buildSimpleProgramEntry
...
+                if (sourceMethod.IsAbstract || sourceMethod.SynthesizesLoweredBoundBody)
+                {
+                    continue;
+                }
+
+                yield return sourceMethod;
+            }
+        }
+
+        internal ImmutableArray<Symbol> GetMembersToMatchAgainsDeclarationSpan()
```
Comment (jjonescz) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:3250]: >Agains [](http://example.com/codeflow?start=57&length=6)

Against

File: src/EditorFeatures/CSharpTest/CodeActions/ExtractMethod/ExtractLocalFunctionTests.cs [hunk 3]
```diff
@@ -1680,9 +1680,9 @@ class C
     public static void Main(string[] args)
     {
         bool Local() => args == null;
-        {|Rename:NewMethod|}(args);
```
Comment (jjonescz) [inline src/EditorFeatures/CSharpTest/CodeActions/ExtractMethod/ExtractLocalFunctionTests.cs:1683]: >args [](http://example.com/codeflow?start=29&length=4)

Why this change? Has it stopped working?

Comment (AlekseyTs) [reply src/EditorFeatures/CSharpTest/CodeActions/ExtractMethod/ExtractLocalFunctionTests.cs:1683]: > Why this change? Has it stopped working?

I think this is a change to the better. The refactoring no longer adds an unused parameter. This is due to an improvement in region analysis around parameters of enclosing methods.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedPrimaryConstructorParameterBackingFieldSymbol.cs (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedPrimaryConstructorParameterBackingFieldSymbol.cs:38]: > Should this return the ParameterSymbol? 

> Note, the set of possible associated symbols might be expanded in the future to reflect changes in the languages.

The set might be expanded, but doesn't have to be expanded. We will do that once we will have a compelling scenario. At the moment we do not have one.

**Comment by AlekseyTs** [comment]
@cston, @jjonescz, @dotnet/roslyn-compiler Please review.

**Comment by AlekseyTs** [comment]
@cston, @jjonescz, @dotnet/roslyn-compiler Please review.

**Review by jjonescz — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs [hunk 4]
```diff
@@ -542,19 +542,50 @@ private SourceEventSymbol GetEventSymbol(EventDeclarationSyntax eventDeclaration
 
             private Symbol GetMemberSymbol(string memberName, TextSpan memberSpan, NamedTypeSymbol container, SymbolKind kind)
             {
-                // return container.GetMembers(methodSyntax.Identifier.ValueText).OfType<SourceMethodSymbol>().Single(m => m.Locations.Any(l => l.SourceTree == tree && methodSyntax.Span.Contains(l.SourceSpan)));
-                foreach (Symbol sym in container.GetMembers(memberName))
+                Debug.Assert(kind is SymbolKind.Method or SymbolKind.Property or SymbolKind.Event);
+
+                if (container is SourceMemberContainerTypeSymbol { PrimaryConstructor: not null } sourceMemberContainerTypeSymbol)
+                {
+                    foreach (Symbol sym in sourceMemberContainerTypeSymbol.GetMembersToMatchAgainstDeclarationSpan())
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs:549]: >sourceMemberContainerTypeSymbol.GetMembersToMatchAgainstDeclarationSpan() [](http://example.com/codeflow?start=43&length=73)

Why use `GetMembersToMatchAgainstDeclarationSpan()` rather than using `GetMembers(memberName)`? Is it to avoid cycles?

File: src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs [hunk 5]
```diff
@@ -333,17 +336,27 @@ protected override ImmutableArray<PendingBranch> Scan(ref bool badRegion)
...
+                        {
+                            NoteWrite(parameter, value: null, read: true);
+                        }
+
+                        CurrentSymbol = save;
                     }
-                }
+                    break;
+
+                case (FieldSymbol or PropertySymbol) and { IsStatic: false, ContainingSymbol: SourceMemberContainerTypeSymbol { PrimaryConstructor: { } primaryConstructor } }:
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs:357]: >case [](http://example.com/codeflow?start=16&length=4)

Is this case doing the same work as the case above? Could the two cases be combined?

File: src/Compilers/CSharp/Portable/Binder/Binder.CapturedParametersFinder.cs [hunk 6]
```diff
@@ -80,7 +80,7 @@ private CapturedParametersFinder(SynthesizedPrimaryConstructor primaryConstructo
                     string name = "<" + parameter.Name + ">PC__BackingField";
```
Comment (tmat) [inline src/Compilers/CSharp/Portable/Binder/Binder.CapturedParametersFinder.cs:80]: >PC__BackingField [](http://example.com/codeflow?start=59&length=16)

I know this has PROTOTYPE on it but for a follow up: we should define GeneratedNameKind entry (e..g `P` looks free to use) and add a helper to GeneratedNames that produces the name.

I'd also just keep the name as short as possible to avoid unnecessary metadata bloat. Could be just `<name>P`.

File: src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs (hunk 4 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/BinderFactory.BinderFactoryVisitor.cs:549]: > Why use ```GetMembersToMatchAgainstDeclarationSpan()``` rather than using GetMembers(memberName)? Is it to avoid cycles?

Yes, ```GetMembers``` includes backing fields, but we need binders to figure them out - cycle.

File: src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs (hunk 5 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs:357]: > Is this case doing the same work as the case above? 

No, they are doing different things. Initializers are considered to be inside the primary constructor. Methods are not.

File: src/Compilers/CSharp/Portable/Binder/Binder.CapturedParametersFinder.cs (hunk 6 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder.CapturedParametersFinder.cs:80]: I will capture these recommendations in comment. Likely in one of the next PRs

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 7]
```diff
@@ -3222,6 +3199,128 @@ ImmutableArray<SynthesizedSimpleProgramEntryPointSymbol> buildSimpleProgramEntry
...
+
+        internal ImmutableArray<Symbol> GetCandidateMembersForLookup(string name)
+        {
+            if (this is { IsRecord: true } or { IsRecordStruct: true } ||
+                this.state.HasComplete(CompletionPart.Members))
+            {
+                return GetMembers(name);
+            }
+
+            ImmutableArray<Symbol> types = GetTypeMembers(name).Cast<NamedTypeSymbol, Symbol>();
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:3274]: >types [](http://example.com/codeflow?start=35&length=5)

Can this local and the call to `GetTypeMembers(name)` be moved after the `if (primaryConstructor.ParameterCount == 0) { ... }` check below?

Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:3274]: If not, perhaps because this call populates the members lists below, consider adding a comment.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:3274]: > Can this local and the call to ```GetTypeMembers(name)``` be moved after the ```if (primaryConstructor.ParameterCount == 0) { ... } check below?```

I'll move it.

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs [hunk 8]
```diff
@@ -496,6 +496,8 @@ BoundExpression makeConditionalAccess(BoundExpression receiver)
 
             static bool isFieldOfMarshalByRef(BoundExpression expr, CSharpCompilation compilation)
             {
+                Debug.Assert(!IsCapturedPrimaryConstructorParameter(expr));
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs:499]: >Debug.Assert(!IsCapturedPrimaryConstructorParameter(expr)) [](http://example.com/codeflow?start=16&length=58)

How do we know the expression is not a primary constructor parameter?

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_UnaryOperator.cs [hunk 9]
```diff
@@ -485,6 +485,7 @@ private static bool IsIndirectOrInstanceField(BoundExpression expression)
                     return ((BoundLocal)expression).LocalSymbol.RefKind != RefKind.None;
 
                 case BoundKind.Parameter:
+                    Debug.Assert(!IsCapturedPrimaryConstructorParameter(expression));
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_UnaryOperator.cs:488]: >Debug.Assert(!IsCapturedPrimaryConstructorParameter(expression)); [](http://example.com/codeflow?start=20&length=65)

How do we know the expression is not a primary constructor parameter?

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs (hunk 8 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs:499]: > How do we know the expression is not a primary constructor parameter?

The ```expr``` is supposed to be lowered before we get here. Captured parameters are replaced with fields during lowering.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_UnaryOperator.cs (hunk 9 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_UnaryOperator.cs:488]: > How do we know the expression is not a primary constructor parameter?

The ```expression``` is supposed to be lowered before we get here. Captured parameters are replaced with fields during lowering.
